### PR TITLE
Skip unavailable storages in scanner

### PR DESCRIPTION
**Pull request by icewind1991**
Instead of crashing we print an errors message and log it.

cc @PVince81 @Xenopathic

### CODE REVIEW

File: lib/private/files/utils/scanner.php [hunk 1]
```diff
@@ -161,7 +170,12 @@ public function scan($dir = '') {
 			if (!$isDbLocking) {
 				$this->db->beginTransaction();
 			}
-			$scanner->scan($relativePath, \OC\Files\Cache\Scanner::SCAN_RECURSIVE, \OC\Files\Cache\Scanner::REUSE_ETAG | \OC\Files\Cache\Scanner::REUSE_SIZE);
+			try {
+				$scanner->scan($relativePath, \OC\Files\Cache\Scanner::SCAN_RECURSIVE, \OC\Files\Cache\Scanner::REUSE_ETAG | \OC\Files\Cache\Scanner::REUSE_SIZE);
+			} catch (StorageNotAvailableException $e) {
+				$this->logger->error('Storage ' . $storage->getId() . ' not available');
```
Comment (MorrisJobke) [inline lib/private/files/utils/scanner.php]: Can we log the full exception here? `logException()`

Comment (icewind1991) [inline lib/private/files/utils/scanner.php]: done

File: lib/private/files/utils/scanner.php [hunk 2]
```diff
@@ -161,7 +170,13 @@ public function scan($dir = '') {
 			if (!$isDbLocking) {
 				$this->db->beginTransaction();
 			}
-			$scanner->scan($relativePath, \OC\Files\Cache\Scanner::SCAN_RECURSIVE, \OC\Files\Cache\Scanner::REUSE_ETAG | \OC\Files\Cache\Scanner::REUSE_SIZE);
+			try {
+				$scanner->scan($relativePath, \OC\Files\Cache\Scanner::SCAN_RECURSIVE, \OC\Files\Cache\Scanner::REUSE_ETAG | \OC\Files\Cache\Scanner::REUSE_SIZE);
+			} catch (StorageNotAvailableException $e) {
```
Comment (PVince81) [inline lib/private/files/utils/scanner.php]: Is it correct to also skip files in "regular scanner" mode, the one that goes to external storage ? Or should we limit it to when run on the CLI?

Comment (icewind1991) [inline lib/private/files/utils/scanner.php]: Correct for both imo

Comment (PVince81) [inline lib/private/files/utils/scanner.php]: You're right. I mixed it up with getDirectoryContent where we cannot affort to remove result entries if they are not available. But this here is the scanner.

**Comment by PVince81** [comment]
Makes sense :+1:

**Comment by PVince81** [comment]
Needs a second reviewer @schiesbn @nickvergessen @DeepDiver1975

**Comment by LukasReschke** [comment]
:+1:

**Comment by PVince81** [comment]
@karlitschek @cmonteroluque backport for 8.2.3 and 8.1.6 ?

**Comment by karlitschek** [comment]
makes sense. please backport :+1:

**Comment by MorrisJobke** [comment]
@icewind1991 Can you open the backport PRs? Thanks

**Comment by icewind1991** [comment]
8.1: #21313
8.1: #21314

**Pull request merged**